### PR TITLE
Add optional write completion handler for emit's

### LINKED ISSUE #1096: Feature Request: completion handler for writes
We've hit a case where we need to know when a write to the underlying socket has been completed. We're sending quite a large amount of latency sensitive data, such that the network can sometimes not keep up. It would be really helpful to be able to monitor when a write actually completes so we can scale back writes when the network slows down. The underlying Starscream WebSokcet class already provides this functionality via a completion handler called on write completion. Ideally this would be exposed up through socket.io as well.

We've considered using the ack mechanism from the server that's already supported by this library, but the extra latency that the network call introduces is unacceptable for our use case. 

Is there a way to achieve this with the current APIs? And if not would you accept a PR adding this functionality?

### PR DESCRIPTION
**Pull request by headlessme**
A first draft to add an optional completion handler for calls to emit, see #1096 for reasoning. I'm not a Swift expert so I'm quite likely doing something dumb here :)

Thanks to @AgentFeeble for providing a great starting point. 

Fixes #1096

### CODE REVIEW

**Comment by nuclearace** [comment]
This looks pretty good! Can you open this against the development branch?

**Comment by headlessme** [comment]
Great! I've changed the base branch

**Comment by nuclearace** [comment]
Excellent, one thought I have though: Would it make sense to wrap the user provided callback with one that wraps and dispatches on the `handleQueue`? If we don't do that we'd need to document on the method that the provided closure won't be executed on the handle queue.

**Comment by nuclearace** [comment]
Something maybe like

```swift
{[weak self] in 
  guard let this = self else { return }
  this.handleQueue.async { /* call callback */ }
}

**Comment by headlessme** [comment]
Good idea – added!

File: Source/SocketIO/Client/SocketIOClient.swift [hunk 1]
```diff
@@ -295,8 +295,14 @@ open class SocketIOClient : NSObject, SocketIOClientSpec {
     }
 
     func emit(_ data: [Any], ack: Int? = nil, binary: Bool = true, isAck: Bool = false, completion: (() -> ())? = nil) {
+        // wrap the completion handler so it always runs async via handlerQueue
+        let wrappedCompletion = {[weak self, completion] in
```
Comment (nuclearace) [inline Source/SocketIO/Client/SocketIOClient.swift]: You shouldn't need to put completion in the capture list.

Comment (headlessme) [reply Source/SocketIO/Client/SocketIOClient.swift]: Fixed

File: Source/SocketIO/Client/SocketIOClient.swift [hunk 2]
```diff
@@ -228,7 +228,17 @@ open class SocketIOClient : NSObject, SocketIOClientSpec {
...
+        emit([event] + items, completion: {})
+    }
+    
+    /// Same as emit, but meant for Objective-C
+    ///
+    /// - parameter event: The event to send.
+    /// - parameter items: The items to send with this event. Send an empty array to send no data.
+    /// - parameter completion: Callback called on transport write completion.
+    @objc
+    open func emit(_ event: String, with items: [Any], completion: @escaping () -> ()) {
```
[the remark this answers is not in the document]
Comment (headlessme) [reply Source/SocketIO/Client/SocketIOClient.swift]: Added a method for Swift. I also added the method signature to the SocketIOClientSpec, but not sure if that's needed?

File: Source/SocketIO/Client/SocketIOClient.swift [hunk 3]
```diff
@@ -213,22 +213,44 @@ open class SocketIOClient : NSObject, SocketIOClientSpec {
...
+    /// Send an event to the server, with optional data items and write completion handler.
+    ///
+    /// If an error occurs trying to transform `items` into their socket representation, a `SocketClientEvent.error`
+    /// will be emitted. The structure of the error data is `[eventName, items, theError]`
+    ///
+    /// - parameter event: The event to send.
+    /// - parameter items: The items to send with this event. May be left out.
+    /// - parameter completion: Callback called on transport write completion.
+    open func emit(_ event: String, _ items: SocketData..., completion: @escaping () -> ())  {
+        emit([event] + items, completion: completion)
```
Comment (nuclearace) [inline Source/SocketIO/Client/SocketIOClient.swift]: This needs to do the transformation from SocketData to something representable in JSON. See the other swift-centric methods.

Comment (headlessme) [reply Source/SocketIO/Client/SocketIOClient.swift]: Ah yea of course, added.

**Comment by headlessme** [comment]
Great! Thanks for the quick review turnaround 👍

**Pull request merged**